### PR TITLE
Minor: Refactor memory size estimation for HashTable

### LINKED ISSUE #8764: Refactor size estimation of Hashset into a function
maybe we can (as a follow on PR) put this logic into its own function (with comments) as estimating the size of hashbrown hashtables is likely to come up again

_Originally posted by @alamb in https://github.com/apache/arrow-datafusion/pull/8721#discussion_r1442010934_

### PR DESCRIPTION
**Pull request by marvinlanhenke**
## Which issue does this PR close?

Closes #8764.

## Rationale for this change

As stated in the issue [here](https://github.com/apache/datafusion/issues/8764#issuecomment-2141501909) the goal is to consolidate estimation logic into a single function. This allows for better documentation as well as better maintainability.
Also there was an [issue ](https://github.com/apache/datafusion/issues/8764#issuecomment-2143467627)with the existing implementation that could still overflow if the `checked_mul` was capped at usize::MAX. I tried to fix this by checking for overflows and returning usize::MAX as a 'max cap'.

## What changes are included in this PR?

- extract logic into single function
- fixed missing overflow checks
- moved logic into datfusion common
- used the new function in `count_distinct/native.rs` and `hash_join.rs`

## Are these changes tested?

Yes.

## Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by marvinlanhenke** [comment]
@alamb @yyy1000 PTAL

**Review by yyy1000 — COMMENTED** [review]
I see it's good abstraction.

My only concern is, it seems that the original goal was kind of ideal, which just pass the hashtable and get the estimate size. But now before using this function, user have to calculate the fixed_size themself, and also the calculation is not always the same.

Though the abstraction is not pretty enough, I think given the current code this PR is still an improvement. :)

**Comment by marvinlanhenke** [comment]
> I see it's good abstraction.
> 
> My only concern is, it seems that the original goal was kind of ideal, which just pass the hashtable and get the estimate size. But now before using this function, user have to calculate the fixed_size themself, and also the calculation is not always the same.
> 
> Though the abstraction is not pretty enough, I think given the current code this PR is still an improvement. :)

Yes, I totally agree with you on this. However, as discussed in the issue, it still might be worth it due to the improved documentation.

File: datafusion/common/src/utils/memory.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module provides a function to estimate the memory size of a HashTable prior to alloaction
+
+/// Estimates the memory size required for a hash table prior to allocation.
+///
+/// # Parameters
+/// - `num_elements`: The number of elements expected in the hash table.
+/// - `fixed_size`: A fixed overhead size associated with the collection
```
Comment (yyy1000) [inline datafusion/common/src/utils/memory.rs:24]: Could we add one or two examples of how could we calculate the `fixed_size`, because this item is kind of abstract . For example, it's the size of all values plus itself for a `Accumulator`

**Comment by yyy1000** [comment]
> However, as discussed in the issue, it still might be worth it due to the improved documentation.

Agree! I left a comment which may improve but it's not necessary. Thanks!

File: datafusion/common/src/utils/memory.rs (hunk 1 above)
Comment (marvinlanhenke) [reply datafusion/common/src/utils/memory.rs:24]: I added both examples

Comment (yyy1000) [reply datafusion/common/src/utils/memory.rs:24]: That's cool and LGTM.

File: datafusion/common/src/utils/memory.rs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+///   - The size of each entry (`T`) multiplied by the estimated number of
+///     buckets.
+///   - One byte overhead for each bucket.
+///   - The fixed size overhead of the collection.
+/// - Returns `usize::MAX` if an overflow occurs.
+///
+/// # Examples
+/// ---
+///
+/// ## From within a struct
```
Comment (alamb) [inline datafusion/common/src/utils/memory.rs:41]: ❤️

File: datafusion/common/src/utils/memory.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        let fixed_size = std::mem::size_of::<HashSet<u32>>();
+        // size: 128 = 16 * 4 + 16 + 48
+        let estimated = estimate_memory_size::<u32>(num_elements, fixed_size);
+
+        assert_eq!(estimated, 128);
+    }
+
+    #[test]
+    fn test_estimate_memory_overflow() {
+        let num_elements = usize::MAX;
```
Comment (alamb) [inline datafusion/common/src/utils/memory.rs:117]: 👍

File: datafusion/common/src/utils/memory.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    use std::collections::HashSet;
+
+    use super::estimate_memory_size;
+
+    #[test]
+    fn test_estimate_memory() {
+        let num_elements = 8;
+        // size: 48
+        let fixed_size = std::mem::size_of::<HashSet<u32>>();
+        // size: 128 = 16 * 4 + 16 + 48
```
Comment (alamb) [inline datafusion/common/src/utils/memory.rs:109]: where does the 16 in this equation come from (the num elements is 8)?

File: datafusion/common/src/utils/memory.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            // + 1 byte for each bucket
+            // + fixed size of collection (HashSet/HashTable)
+            std::mem::size_of::<T>()
+                .checked_mul(estimated_buckets)
+                .and_then(|size_of_entries| {
+                    estimated_buckets
+                        .checked_add(fixed_size)
+                        .and_then(|overhead| size_of_entries.checked_add(overhead))
+                })
+        })
```
Comment (alamb) [inline datafusion/common/src/utils/memory.rs:94]: I found this formulation very hard to follow due to all the unesting

I wonder if we could use the `?` operator to make it read better


Something like

```rust
    num_elements
        .checked_mul(8)
        .and_then(|estimate| estimate.checked_div(7))
        .map(|buckets| buckets.next_power_of_two())
        .and_then(|estimated_buckets| {
            // + size of entry * number of buckets
            std::mem::size_of::<T>()
                .checked_mul(estimated_buckets)?
                // + 1 byte for each bucket
                .checked_add(estimated_buckets)?
                // + fixed size of collection (HashSet/HashTable)
                .checked_add(fixed_size)
        })
        .unwrap_or(usize::MAX)
```

File: datafusion/common/src/utils/memory.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+    #[test]
+    fn test_estimate_memory() {
+        let num_elements = 8;
+        // size: 48
+        let fixed_size = std::mem::size_of::<HashSet<u32>>();
+        // size: 128 = 16 * 4 + 16 + 48
+        let estimated = estimate_memory_size::<u32>(num_elements, fixed_size);
+
+        assert_eq!(estimated, 128);
```
Comment (alamb) [inline datafusion/common/src/utils/memory.rs:112]: could you please also add a test with the same fixed_size but a different num_elements (e.g. 40) and show what that value was?

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 7]
```diff
@@ -875,23 +875,12 @@ async fn collect_left_input(
 
     // Estimation of memory size, required for hashtable, prior to allocation.
     // Final result can be verified using `RawTable.allocation_info()`
-    //
-    // For majority of cases hashbrown overestimates buckets qty to keep ~1/8 of them empty.
-    // This formula leads to overallocation for small tables (< 8 elements) but fine overall.
-    let estimated_buckets = (num_rows.checked_mul(8).ok_or_else(|| {
```
Comment (alamb) [inline datafusion/physical-plan/src/joins/hash_join.rs:881]: Another alternative is to make `estimate_memory_size` return `Result<usize>` rather than `usize` 

In general if the cost model overflows that is probably something that would be good to flag as an error rather than continuing calculation

**Review by alamb — COMMENTED** [review]
Thank you so much @marvinlanhenke  -- this looks like a great step forward in readability 🙏 

Thank you @yyy1000  for your review

I left some suggestions / comments. Let me know what you think

File: datafusion/common/src/utils/memory.rs (hunk 4 above)
Comment (marvinlanhenke) [reply datafusion/common/src/utils/memory.rs:109]: Ah sorry it's not well documented...16 is the number of estimated buckets

File: datafusion/physical-plan/src/joins/hash_join.rs (hunk 7 above)
Comment (marvinlanhenke) [reply datafusion/physical-plan/src/joins/hash_join.rs:881]: Yes, I think this would be the cleanest solution. I wrote a comment in the issue, since this would require us to change the Accumulator trait; and I'm not sure we want to do this, for a probably rare edge-case?

File: datafusion/common/src/utils/memory.rs (hunk 5 above)
Comment (marvinlanhenke) [reply datafusion/common/src/utils/memory.rs:94]: Wouldn't this also require us to change the fn to Result<usize>, which leads to a change in the Accumulator trait.

Comment (alamb) [reply datafusion/common/src/utils/memory.rs:94]: I don't think so (I actually tried that code above and it works fine)

In rust `?` can also be used in functions that return `Option`  (to return `None`)

File: datafusion/physical-plan/src/joins/hash_join.rs (hunk 7 above)
Comment (alamb) [reply datafusion/physical-plan/src/joins/hash_join.rs:881]: Perhaps we can simply ignore the error in the accumulator (e.g. do the unwrap in at the callsite) and discuss changing the accumulator trait in some future PR?

Comment (alamb) [reply datafusion/physical-plan/src/joins/hash_join.rs:881]: Sorry -- i missed the [comment](https://github.com/apache/datafusion/issues/8764#issuecomment-2143467627) on the ticket

I think we should change the code to panic on overflow and we can sort out what to do if that is seen in practice

Comment (marvinlanhenke) [reply datafusion/physical-plan/src/joins/hash_join.rs:881]: so you mean unwrap() the error at callsite (the accumulator)?

**Comment by marvinlanhenke** [comment]
@alamb 
I made the changes based on your comments. Thanks for the comprehensive review. 🚀 

I changed `fn estimate_memory_size(..)` to return `Result<usize>` which should set us up for any possible future changes in the Accumulator trait; for now I simply unwrap in `fn size()` and panic.

File: datafusion/common/src/utils/memory.rs (hunk 6 above)
Comment (marvinlanhenke) [reply datafusion/common/src/utils/memory.rs:116]: - added num_elements = 40.
- made the comments more explicit

File: datafusion/common/src/utils/memory.rs (hunk 5 above)
Comment (marvinlanhenke) [reply datafusion/common/src/utils/memory.rs]: thanks for the suggestion; this looks way nicer 🚀

File: datafusion/common/src/utils/memory.rs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+pub fn estimate_memory_size<T>(num_elements: usize, fixed_size: usize) -> Result<usize> {
+    // For the majority of cases hashbrown overestimates the bucket quantity
+    // to keep ~1/8 of them empty. We take this factor into account by
+    // multiplying the number of elements with a fixed ratio of 8/7 (~1.14).
+    // This formula leads to overallocation for small tables (< 8 elements)
+    // but should be fine overall.
+    num_elements
+        .checked_mul(8)
+        .and_then(|overestimate| {
+            let estimated_buckets = (overestimate / 7).next_power_of_two();
```
Comment (alamb) [inline datafusion/common/src/utils/memory.rs:86]: 👍

File: datafusion/physical-expr/src/aggregate/count_distinct/native.rs [hunk 9]
```diff
@@ -115,18 +116,11 @@ where
...
-        // + fixed size of HashSet
-        std::mem::size_of_val(self)
-            + std::mem::size_of::<T::Native>() * estimated_buckets
-            + estimated_buckets
-            + std::mem::size_of_val(&self.values)
+        let num_elements = self.values.len();
+        let fixed_size =
+            std::mem::size_of_val(self) + std::mem::size_of_val(&self.values);
+
+        estimate_memory_size::<T::Native>(num_elements, fixed_size).unwrap()
```
Comment (alamb) [inline datafusion/physical-expr/src/aggregate/count_distinct/native.rs:123]: 👍  I think this makes sense to me.

**Review by alamb — APPROVED** [review]
Thank you @marvinlanhenke

**Comment by alamb** [comment]
Thanks again @marvinlanhenke
